File: 1112-FindWordsThatCanBeFormedByCharacters/1112-FindWordsThatCanBeFormedByCharacters.py
# Last updated: 6/9/2025, 5:25:50 PM
class Solution:
    def countCharacters(self, words: List[str], chars: str) -> int:
        # check if word in list of words can be made from characters in chars each once 
        # return sum of lengths of all good words 
        # if each letter in word is in chars, remove it from chars 
        # freq map for each word? 
        # how many times each character occurs in char
        # key value, for each letter in word, check that letter, value =< charsmap

        # length 
        length = 0 
        # freq map 
        freq = defaultdict(int)

        # map chars 
        for c in chars: 
            freq[c] += 1
        
        # check each word
        for word in words: 
            # check if char in word shows up more than in char - invalid 
            for char in word: 
                if word.count(char) > freq[char]:
                    break
            # if char in word shows up less or eq to char in given 
            else:
                #add on length
                length += len(word)
        
        return length
        
    

                 
        

File: 1112-FindWordsThatCanBeFormedByCharacters/1112-FindWordsThatCanBeFormedByCharacters.java
// Last updated: 11/1/2025, 4:59:15 PM
class Solution {
    public int countCharacters(String[] words, String chars) {
        // U: validate if word in words can be made from string chars then sum lengths of validated words
        // M: for loop, len(word), letter in string
        // P: 
        // for each word in words, 
        //  for each letter in word, 
        //     if letter not in chars, move on to next word
        //  end of word & no letters not in chars, add to sum 

        // dictionary/ hashmap 
        // map for each word
        // map for chars 
        // check if each key, value is either key, value or key, value < chars key, value

        // create freq map for chars
        Map<Character, Integer> charFreq = new HashMap<>();

        for (char c: chars.toCharArray()) {
            charFreq.put(c, charFreq.getOrDefault(c, 0) + 1);
        }

        //counter for sum
        int totalLen = 0;

        // for each word in words
        for (String word: words) {
            Map<Character, Integer> wordFreq = new HashMap<>(); 
            for (char c: word.toCharArray()) {
                wordFreq.put(c, wordFreq.getOrDefault(c, 0) +1);
            }
            // check if word can be formed by chars
            boolean valid = true;
            for (char c: word.toCharArray()) {
                if (wordFreq.get(c) > charFreq.getOrDefault(c, 0)) {
                    valid = false; 
                    break;
                }
            }
            
            if (valid) {
            totalLen += word.length();
              }
        }
        

        return totalLen;
    }
}